cdap-data-fabric/src/main/java/co/cask/cdap/data/view/MDSViewStore.java

```diff
@@ -85,70 +107,60 @@ public Boolean apply(ViewMetadataStoreDataset mds) throws Exception {
 
   @Override
   public boolean exists(final Id.Stream.View viewId) {
-    return execute(new TransactionExecutor.Function<ViewMetadataStoreDataset, Boolean>() {
+    return execute(new TxCallable<Boolean>() {
       @Override
-      public Boolean apply(ViewMetadataStoreDataset mds) throws Exception {
-        return mds.exists(getKey(viewId));
+      public Boolean call(DatasetContext context) throws Exception {
+        return getViewDataset(context).exists(getKey(viewId));
       }
     });
   }
 
   @Override
   public void delete(final Id.Stream.View viewId) throws NotFoundException {
-    boolean notFound = execute(new TransactionExecutor.Function<ViewMetadataStoreDataset, Boolean>() {
-      @Override
-      public Boolean apply(ViewMetadataStoreDataset mds) throws Exception {
-        if (!mds.exists(getKey(viewId))) {
-          return true;
+    try {
+      transactional.execute(new TxRunnable() {
+        @Override
+        public void run(DatasetContext context) throws Exception {
+          ViewMetadataStoreDataset mds = getViewDataset(context);
+          MDSKey key = getKey(viewId);
+          if (!mds.exists(key)) {
+            throw new NotFoundException(viewId);
+          }
+          mds.deleteAll(key);
         }
-        mds.deleteAll(getKey(viewId));
-        return false;
-      }
-    });
-
-    if (notFound) {
-      throw new NotFoundException(viewId);
+      });
+    } catch (TransactionFailureException e) {
+      throw Transactions.propagate(e, NotFoundException.class);
     }
   }
 
   @Override
   public List<Id.Stream.View> list(final Id.Stream streamId) {
-    List<StreamViewEntry> entries = execute(
-      new TransactionExecutor.Function<ViewMetadataStoreDataset, List<StreamViewEntry>>() {
-        @Override
-        public List<StreamViewEntry> apply(ViewMetadataStoreDataset mds) throws Exception {
-          return Objects.firstNonNull(
-            mds.<StreamViewEntry>list(getKey(streamId), StreamViewEntry.class),
-            ImmutableList.<StreamViewEntry>of());
-        }
-      });
-
-    ImmutableList.Builder<Id.Stream.View> builder = ImmutableList.builder();
-    builder.addAll(Collections2.transform(entries, new Function<StreamViewEntry, Id.Stream.View>() {
+    return execute(new TxCallable<List<Id.Stream.View>>() {
       @Override
-      public Id.Stream.View apply(StreamViewEntry input) {
-        return input.getId();
+      public List<Id.Stream.View> call(DatasetContext context) throws Exception {
+        List<StreamViewEntry> views = getViewDataset(context).list(getKey(streamId), StreamViewEntry.class);
+        return ImmutableList.copyOf(Lists.transform(views, VIEW_ENTRY_TO_ID));
```

Comment:
The main difference is whether the `VIEW_ENTRY_TO_ID.apply()` will get triggered once or more than one. If we return the transformed list, then every access to the list entry will have the conversion function triggered. I think either copying or having the trigger method called multiple times isn't really a big deal in this context (the old code was copying, so returning an immutable copy is a safer bet for this change).